compiler/code-gen/declarations.cpp

```diff
@@ -1141,16 +1141,26 @@ void ClassMembersDefinition::compile_msgpack_serialize(CodeGenerator &W, ClassPt
   //   packer.pack(field_1);
   //   ...
   //}
-  std::vector<std::string> body;
+  std::list<std::string> body;
   uint16_t cnt_fields = 0;
 
-  klass->members.for_each([&](ClassMemberInstanceField &field) {
-    if (field.serialization_tag != -1) {
-      auto func_name = fmt_format("vk::msgpack::packer_float32_decorator::pack_value{}", field.serialize_as_float32 ? "_float32" : "");
-      body.emplace_back(fmt_format("packer.pack({}); {}(packer, ${});", field.serialization_tag, func_name, field.var->name));
-      cnt_fields += 2;
+  ClassPtr the_klass = klass;
+
+  // Put class' fields along with all parent's fields.
+  while (the_klass) {
+    std::vector<std::string> body_inner;
+    the_klass->members.for_each([&](ClassMemberInstanceField &field) {
+      if (field.serialization_tag != -1) {
+        auto func_name = fmt_format("vk::msgpack::packer_float32_decorator::pack_value{}", field.serialize_as_float32 ? "_float32" : "");
+        body_inner.emplace_back(fmt_format("packer.pack({}); {}(packer, ${});", field.serialization_tag, func_name, field.var->name));
+        cnt_fields += 2;
+      }
+    });
+    for (auto it = body_inner.rbegin(); it != body_inner.rend(); ++it) {
```

Comment:
Is it mandatory to iterate in reverse order here?

Reply:
> Is it mandatory to iterate in reverse order here?

No, it is not mandatory. But the reason to reverse values here is to save natural order of cases in resulting code. Take a look to generated code: cases are ordered naturally same as `@kphp-serialized-field` indices.
Insertion of parent's fields is done in _reverse_ order, that's why values are inserted backwards.

`std::list` (or even `forward_list`) is a good choice if you need to insert in reverse order with further usage in direct order.

Reply:
It's not actually. It requires a memory allocation on each element addition. At the same time `std::deque` allows you to each insert/delete elements from both its ends without needing to allocate memory each time you insert an element.

Also, it's recommended to avoid explicit loops in C++. For example, here you could use, for example, `std::move` algorithm from `<algorithm>`. It makes code easier to read and could also increase performance due to heavy optimization deep inside standard library.

Reply:
> It's not actually. It requires a memory allocation on each element addition. At the same time `std::deque` allows you to each insert/delete elements from both its ends without needing to allocate memory each time you insert an element.

Sure, you're right, but it is done in a cost of preliminary allocation. According to `std::deque` [documentation](https://en.cppreference.com/w/cpp/container/deque), it is significantly more expensive than vector:

```
On the other hand, deques typically have large minimal memory cost; 
a deque holding just one element has to allocate its full internal array 
(e.g. 8 times the object size on 64-bit libstdc++; 16 times the object size 
or 4096 bytes, whichever is larger, on 64-bit libc++).
```

In fact, we have one-way memory allocation for small amount of fields.

Most (30%) classes has no more than 6 fields, deque usage will be memory overkill.
Also, it should be noted memory is allocated (and immediately deallocated) only during code generation routines on per-class bases in `compile_msgpack_serialization` and `compile_msgpack_deserialization`, so, not think it is a problem to use `list` here. We have about 2000 classes with serialization, not that much.

Reply:
> Also, it's recommended to avoid explicit loops in C++. For example, here you could use, for example, `std::move` algorithm from `<algorithm>`. It makes code easier to read and could also increase performance due to heavy optimization deep inside standard library.

I found no usage of `std::move` from `<algorithm>` in our codebase.
Don't want to be first :)